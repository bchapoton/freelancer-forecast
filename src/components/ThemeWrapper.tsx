import React, { ReactNode } from 'react';
import { createTheme, ThemeProvider } from '@mui/material';

export type ThemeWrapperProps = {
    children: ReactNode;
};

function ThemeWrapper({ children }: ThemeWrapperProps) {
    const theme = createTheme({
        typography: {
            fontSize: 13,
        },
        components: {
            MuiTableRow: {
                styleOverrides: {
                    head: {
                        color: '#495057',
                        backgroundColor: '#eff2f7',
                        fontWeight: 600,
                        borderBottomColor: '#d7dade',
                        borderBottomWidth: '2px',
                        borderBottomStyle: 'solid',
                    },
                    hover: {
                        backgroundColor: '#495057',
                    },
                    root: {
                        '&:last-child td, &:last-child th': { border: 0 },
                    },
                },
            },
            MuiTableCell: {
                styleOverrides: {
                    head: {
                        fontSize: '13px',
                        fontWeight: 600,
                        color: '#495057',
                    },
                    body: {
                        color: '#495057',
                        fontSize: '13px',
                    },
                },
            },
            MuiTableContainer: {
                styleOverrides: {
                    root: {
                        borderRadius: 5,
                    },
                },
            },
        },
    });
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

export default ThemeWrapper;
