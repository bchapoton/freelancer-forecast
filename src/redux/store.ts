import { configureStore } from '@reduxjs/toolkit';
import parametersReducer from './slices/ParametersSlice';
import summariesReducer from './slices/SummariesSlice';
import uiReducer from './slices/UISlice';
import { StorePreloadedState } from '../helpers/TestHelper';

export const store = storeConfigurator();

export function storeConfigurator(preloadedState?: StorePreloadedState) {
    return configureStore({
        reducer: {
            parameters: parametersReducer,
            summaries: summariesReducer,
            ui: uiReducer,
        },
        preloadedState: preloadedState,
    });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
