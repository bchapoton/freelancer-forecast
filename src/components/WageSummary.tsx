import React, { Fragment, useEffect, useState } from 'react';
import { Paper, Table, TableBody, TableCell, TableContainer, TableHead, TableRow, Typography } from '@mui/material';
import Euro from './Euro';
import parametersService, { YearSummary } from '../services/ParametersService';
import { useAppSelector } from '../redux/hooks';
import { selectIncomeTaxesSummary, selectYearSummary } from '../redux/slices/FinancialSlice';
import { ParametersState, selectParameters } from '../redux/slices/ParametersSlice';
import { IncomeTaxesSummary } from '../services/TaxesService';

function WageSummary() {
    const yearSummary: YearSummary = useAppSelector(selectYearSummary);
    const parameters: ParametersState = useAppSelector(selectParameters);
    const incomeTaxesResult: IncomeTaxesSummary = useAppSelector(selectIncomeTaxesSummary);

    const [values, setValues] = useState<ValuesState>({
        wagesBeforeTaxesPerYear: 0,
        savingPerYear: 0,
        incomeTaxesPerYear: 0,
        wagesAfterTaxesPerYear: 0,
    });

    useEffect(() => {
        const saving = parametersService.calculateSaving(
            yearSummary.totals.profits - incomeTaxesResult.total,
            parameters.saving,
        );
        setValues({
            wagesBeforeTaxesPerYear: yearSummary.totals.profits - saving,
            incomeTaxesPerYear: incomeTaxesResult.total,
            wagesAfterTaxesPerYear: yearSummary.totals.profits - incomeTaxesResult.total - saving,
            savingPerYear: saving,
        });
    }, [yearSummary, incomeTaxesResult, parameters]);

    return (
        <Fragment>
            <Typography gutterBottom variant="h5" component="div">
                Salaires
            </Typography>
            <Paper>
                <TableContainer>
                    <Table size="small">
                        <TableHead>
                            <TableRow>
                                <TableCell>&nbsp;</TableCell>
                                <TableCell align="right">Mensuel lissé sur 12 mois</TableCell>
                                <TableCell align="right">Annuel</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            <TableRow hover={true}>
                                <TableCell component="th" scope="row">
                                    Salaire avant impôts sur le revenu
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.wagesBeforeTaxesPerYear / 12}</Euro>
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.wagesBeforeTaxesPerYear}</Euro>
                                </TableCell>
                            </TableRow>
                            <TableRow hover={true}>
                                <TableCell component="th" scope="row">
                                    Impôts sur le revenu
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.incomeTaxesPerYear / 12}</Euro>
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.incomeTaxesPerYear}</Euro>
                                </TableCell>
                            </TableRow>
                            <TableRow hover={true}>
                                <TableCell component="th" scope="row">
                                    Salaire après impôts sur le revenu
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.wagesAfterTaxesPerYear / 12}</Euro>
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.wagesAfterTaxesPerYear}</Euro>
                                </TableCell>
                            </TableRow>
                            <TableRow hover={true}>
                                <TableCell component="th" scope="row">
                                    Epargne entreprise
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.savingPerYear / 12}</Euro>
                                </TableCell>
                                <TableCell align="right">
                                    <Euro>{values.savingPerYear}</Euro>
                                </TableCell>
                            </TableRow>
                        </TableBody>
                    </Table>
                </TableContainer>
            </Paper>
        </Fragment>
    );
}

export default WageSummary;

type ValuesState = {
    wagesBeforeTaxesPerYear: number;
    incomeTaxesPerYear: number;
    wagesAfterTaxesPerYear: number;
    savingPerYear: number;
};
