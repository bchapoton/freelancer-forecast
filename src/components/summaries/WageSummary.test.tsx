import React from 'react';
import renderer from 'react-test-renderer';
import { store } from '../../redux/store';
import { Provider } from 'react-redux';
import ThemeWrapper from '../ThemeWrapper';
import WageSummary from './WageSummary';

it('Wage summary renders correctly', () => {
    const tree = renderer
        .create(
            <Provider store={store}>
                <ThemeWrapper>
                    <WageSummary />
                </ThemeWrapper>
            </Provider>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
