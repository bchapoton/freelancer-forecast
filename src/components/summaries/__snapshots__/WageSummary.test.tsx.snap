// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wage summary renders correctly 1`] = `
Array [
  <div
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1pzmhef-MuiTypography-root"
  >
    Salaires
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
  >
    <div
      className="MuiTableContainer-root css-1lqbn85-MuiTableContainer-root"
    >
      <table
        className="MuiTable-root css-1m14wjh-MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head css-xev7fm-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-t8mu5n-MuiTableCell-root"
              scope="col"
            >
               
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-abrxy2-MuiTableCell-root"
              scope="col"
            >
              Mensuel lissé sur 12 mois
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-abrxy2-MuiTableCell-root"
              scope="col"
            >
              Annuel
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-hover css-d9l2c8-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-9xppz4-MuiTableCell-root"
              scope="row"
            >
              Salaire avant impôts sur le revenu
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root MuiTableRow-hover css-d9l2c8-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-9xppz4-MuiTableCell-root"
              scope="row"
            >
              Impôts sur le revenu
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root MuiTableRow-hover css-d9l2c8-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-9xppz4-MuiTableCell-root"
              scope="row"
            >
              Salaire après impôts sur le revenu
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
          </tr>
          <tr
            className="MuiTableRow-root MuiTableRow-hover css-d9l2c8-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-9xppz4-MuiTableCell-root"
              scope="row"
            >
              Epargne entreprise
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-egzx5e-MuiTableCell-root"
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0 €
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
]
`;
